### ISSUE TITLE
DNF5 always removes partially downloaded files when interrupted, so you cannot resume downloading them

### ISSUE THREAD

**Issue by hedayat** [rpm-software-management/librepo#293]
I tried both running `dnf5 install ...` and `dnf5 download ...` in Fedora 39, and interrupting them in the middle of downloading a large package. Upon interruption, the partially downloaded file is removed. Hence, you cannot resume downloading the package. I tried also adding `keepcache` option, but it has no effect.

Effectively, DNF5 doesn't have support for resume functionality right now.

**Comment by jan-kolarik** [comment]
Hi, I've been testing the described issue on my local machine, and it appears to be related to the librepo component handling package downloads. Therefore, I've transferred the problem there.

The issue lies in how `librepo` processes interruptions. When a `SIGTERM` is sent to the DNF process, the partially downloaded file is retained and resumed upon the next download attempt. However, when a `SIGINT` (<kbd>CTRL</kbd>+<kbd>C</kbd>) is sent, the file is discarded, resulting in the following error:

```
>>> Not finished - interrupted by error: Interrupted by signalFailed to download packages
 Librepo error: Interrupted by a SIGINT signal
```

Upon comparing this behavior to DNF4, I observed that it behaves similarly, just with different output.

**Comment by jan-kolarik** [comment]
I wasn't investigating deeply why the file is removed when `SIGINT` is used, but it seems that the issue may be related to this section: https://github.com/rpm-software-management/librepo/blob/master/librepo/downloader.c#L2817.

**Comment by hedayat** [comment]
> Upon comparing this behavior to DNF4, I observed that it behaves similarly, just with different output.

TBH, I hope DNF5 doesn't try to replicate bad behaviors from DNF4 :P Although, I agree it depends on how you interpret Ctrl+C: Cancel vs Interrupt.

But, if the current behavior is retains, so we should either terminate DNF5 by other means or hope it'll terminate by itself upon timeout/...? 

And, does it mean that if  DNF has downloaded a number of packages and you press Ctrl+C, it'll remove all of them? Not sure that's what a user expects :thinking:

**Comment by jan-kolarik** [comment]
> TBH, I hope DNF5 doesn't try to replicate bad behaviors from DNF4 :P Although, I agree it depends on how you interpret Ctrl+C: Cancel vs Interrupt.

> But, if the current behavior is retains, so we should either terminate DNF5 by other means or hope it'll terminate by itself upon timeout/...?

I've just done preliminary analysis of the issue, but I don't think that the current behavior is good. It should be fixed.

> And, does it mean that if DNF has downloaded a number of packages and you press Ctrl+C, it'll remove all of them? Not sure that's what a user expects 🤔

No, pressing Ctrl+C while downloading packages will only remove the package currently being downloaded, but not those that have already finished downloading.

**Comment by hedayat** [comment]
> I've just done preliminary analysis of the issue, but I don't think that the current behavior is good. It should be fixed.

Sure, thanks :) 


> No, pressing Ctrl+C while downloading packages will only remove the package currently being downloaded, but not those that have already finished downloading.

Good. Because I thought maybe the decision was made that when you press Ctrl+C, you want to cancel the operation. In that case, other package should have been removed. Since they are not removed, I can safely assume that retaining partial file is also OK.

**Comment by jan-kolarik** [comment]
Another use case that should be covered when fixing this: https://github.com/rpm-software-management/dnf/issues/2051#issuecomment-1946148621.

**Comment by radiumatic** [comment]
I thought this issue was fixed, but apparently not.
Has there been any progress? 
dnf5 successfully switches between mirrors when a timeout occurs, but when "Failure when receiving data from the peer for X" happens, it just wipes the file.

**Comment by slabua** [comment]
Same issue here, still partially downloaded files not resuming.

**Comment by kontura** [comment]
Fedora report: https://bugzilla.redhat.com/show_bug.cgi?id=2397080

**Comment by radiumatic** [comment]
I think the two checks that https://github.com/rpm-software-management/librepo/blob/master/librepo/downloader.c performs for resuming a download and which have a bug that results in this behavior are not needed at all. The checksum will catch any corruption so they can be safely removed. 

Can someone please try this? I can't install all the build dependencies at the moment.

**Comment by radiumatic** [comment]
Two years, three and half a month.

**Comment by m-blaha** [comment]
Yeah sorry for that. Contributions are always highly appreciated though...

Anyway, I investigated the issue and found it more complex than anticipated. Partial download resumption has never actually functioned in librepo. Several scenarios would benefit from this feature:

- Within a single dnf command: after a network failure, the library currently discards already downloaded data and restarts the download from scratch using another mirror. This is clearly an implementation bug, as this functionality was expected to work. (This is what https://github.com/rpm-software-management/librepo/pull/369 is supposed to fix)

- `dnf install <a large rpm>` intentionally interrupted during download: resumption should probably work in this case, even across dnf calls.

- `dnf download <a large rpm>` interrupted: dnf arguably should not leave partially downloaded files behind, at least not without explicit user request (for example, a configuration option like `keep_partial_downloads = true`).

**Comment by radiumatic** [comment]
Do RSTs and EoF errors still cause the file to be wiped?
